core/arch/arm/kernel/early_ta.c

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (c) 2017, Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <initcall.h>
+#include <kernel/early_ta.h>
+#include <kernel/linker.h>
+#include <kernel/user_ta.h>
+#include <stdio.h>
+#include <string.h>
+#include <trace.h>
+#include <util.h>
+#include <zlib.h>
+
+#include "elf_load.h"
+
+struct user_ta_store_handle {
+	const struct early_ta *early_ta;
+	size_t offs;
+	z_stream strm;
+};
+
+#define for_each_early_ta(_ta) \
+	for (_ta = &__rodata_early_ta_start; _ta < &__rodata_early_ta_end; \
+	     _ta = (const struct early_ta *)				   \
+		   ROUNDUP((vaddr_t)_ta + sizeof(*_ta) + _ta->size, 8))
+
+static const struct early_ta *find_early_ta(const TEE_UUID *uuid)
+{
+	const struct early_ta *ta;
+
+	for_each_early_ta(ta)
+		if (!memcmp(&ta->uuid, uuid, sizeof(*uuid)))
+			return ta;
+
+	return NULL;
+}
+
+static void *zalloc(void *opaque __unused, unsigned int items,
+		    unsigned int size)
+{
+	return malloc(items * size);
+}
+
+static void zfree(void *opaque __unused, void *address)
+{
+	free(address);
+}
+
+static bool decompression_init(z_stream *strm,
+			       const struct early_ta *ta)
+{
+	int st;
+
+	strm->next_in = ta->ta;
```

Comment:
Thanks for the hint, I'll fix this whole `const` handling issue by defining `ZLIB_CONST` and removing `-Wno-discarded-qualifiers`.